source/tutorials/module-system/a-basic-module/index.md

```diff
@@ -0,0 +1,108 @@
+# A basic module
+
+What is a module?
+
+* A module is a function that takes an attrset and returns an attrset.
+* It *may* declare options.
+* It *may* define option values.
+* When evaluated, it produces a configuration based on the declarations and definitions.
+
+The format is like so:
+
+```{code-block} nix
+:caption: useless.nix
+{...}: {
+}
+```
+
+This, as the filename suggests, is completely useless.
+It takes no arguments and returns an empty attrset.
+However, it is a valid module.
+Let us add to this to make it a bit more useful.
+
+To define any values, the module system first has to know which ones are allowed.
+This is done by declaring options that specify which values can be set and used elsewhere.
+Options are declared by adding an attribute under the top-level `options` attribute.
+The most general way to declare an option is using `lib.mkOption`.
+
+```{literalinclude} options.nix
+:language: nix
+:caption: options.nix
+```
+
+While many attributes for customizing options are available,
+the most important one is `type`,
+which specifies which values are valid for an option.
+There are several types available under [`lib.types`][option-types-basic] in the Nixpkgs library.
+
+As you can see, we have declared an option `name`.
+We have specificied that the `name` option will be of type `str`, so the module system will expect a string when we set a value.
+
+You may have noticed that we also changed the function arguments.
+Now the module is a function which takes *at least* one argument, `lib`,
+and may accept other arguments (expressed by the ellipsis `...`).
+This will make Nixpkgs library functions available within the function body.
+We needed this to get access to `mkOption` and `types`.
+
+:::{note}
+The ellipsis `...` is necessary because arbitrary arguments can be passed to modules.
+Every module should have this.
+
+The `lib` argument is passed automatically by the module system.
+It is absolutely vital for modules that have option declarations, as you will need `lib` for defining options and their types.
+It is one of several arguments that are automatically provided by the module system.
+The full list of arguments is discussed later.
+:::
+
+Now that we have declared an option, we would naturally want to give it a value.
+Options can be set or *defined* using another top-level attribute, `config`.
+
+```{literalinclude} config.nix
+:language: nix
+:caption: config.nix
+```
+
+Previously, in our option declaration, we created an option, `name`, with a string type.
```

Comment:
It might be worth mentioning that in this case you made a separate file with the config, and that they will all be merged together in the end 🤔